### ISSUE TITLE
Seg fault filling root event

### ISSUE THREAD

**Issue by nickwp** [WCSim/WCSim#463]
While simulating pi+ in WCTE, I am very rarely (simulated 3 million events, saw the seg fault on two indpendent events) getting a segfault when filling the root event. I have not seen any seg faults when simulating 3M muons or 3M electrons.

I expect the limit is passed for some fixed sized array in the output class. I'll try to debug this when I get a chance, but reporting it here in case anyone has a good idea what it might be and could help to diagnose it more quickly.

```
 *** Break *** segmentation violation
 Generating stack trace...
 0x00001470988b4497 in WCSimEventAction::FillRootEventHybrid(int, ntupleStruct const&, G4TrajectoryContainer*, G4TDigiCollection<WCSimWCDigi>*, G4TDigiCollection<WCSimWCDigiTrigger>*, G4String, WCSimRootEvent*, WCSimRootTrigger*) + 0x1047 from /opt/WCSim/build/install/lib/libWCSimCore.so.1.12.9
 0x00001470988c0d1e in WCSimEventAction::EndOfEventAction(G4Event const*) + 0x394e from /opt/WCSim/build/install/lib/libWCSimCore.so.1.12.9
 0x00001470969f60b3 in G4EventManager::DoProcessing(G4Event*) + 0x8f3 from /opt/geant4/lib/libG4event.so
 0x0000147097ddf607 in G4RunManager::DoEventLoop(int, char const*, int) + 0xb7 from /opt/geant4/lib/libG4run.so
 0x0000147097ddd3c2 in G4RunManager::BeamOn(int, char const*, int) + 0x62 from /opt/geant4/lib/libG4run.so
 0x0000147097df59d7 in G4RunMessenger::SetNewValue(G4UIcommand*, G4String) + 0x5c7 from /opt/geant4/lib/libG4run.so
 0x0000147097d636c0 in G4UIcommand::DoIt(G4String) + 0x530 from /opt/geant4/lib/libG4intercoms.so
 0x0000147097d7e978 in G4UImanager::ApplyCommand(char const*) + 0xde8 from /opt/geant4/lib/libG4intercoms.so
 0x0000147097d51351 in G4UIbatch::ExecCommand(G4String const&) + 0x21 from /opt/geant4/lib/libG4intercoms.so
 0x0000147097d53146 in G4UIbatch::SessionStart() + 0x76 from /opt/geant4/lib/libG4intercoms.so
 0x0000147097d7fc86 in G4UImanager::ExecuteMacroFile(char const*) + 0x36 from /opt/geant4/lib/libG4intercoms.so
 0x0000147097d6e0aa in G4UIcontrolMessenger::SetNewValue(G4UIcommand*, G4String) + 0x24a from /opt/geant4/lib/libG4intercoms.so
 0x0000147097d636c0 in G4UIcommand::DoIt(G4String) + 0x530 from /opt/geant4/lib/libG4intercoms.so
 0x0000147097d7e978 in G4UImanager::ApplyCommand(char const*) + 0xde8 from /opt/geant4/lib/libG4intercoms.so
 0x000055a6b489de5e in main + 0x68e from WCSim
 0x0000147097873083 in __libc_start_main + 0xf3 from /usr/lib/x86_64-linux-gnu/libc.so.6
 0x000055a6b489e21e in _start + 0x2e from WCSim
```

**Comment by nickwp** [comment]
To reproduce, try the attached mac file (renamed to .txt. so that github lets me attach it)
I get segfault on event 314
[wcsim_wcte_pgun_pi+_703.txt](https://github.com/user-attachments/files/16321602/wcsim_wcte_pgun_pi%2B_703.txt)

**Comment by kmtsui** [comment]
Any printout before the seg fault?

**Comment by kmtsui** [comment]
Ok I am able to reproduce the seg fault with the WCTE container.

By putting in many verbose, I am able to track down the problem to [`WCSimEventAction::WCSimEventFindStartingVolume()`](https://github.com/WCSim/WCSim/blob/a5c9d56908f39b0daedab41c674730d04ce11771/src/WCSimEventAction.cc#L1021C7-L1021C53), where a vertex outside the detector volume is input into the function. 

A quick fix is to exit the function if `tmpVolume` is invalid.
```
  G4VPhysicalVolume* tmpVolume = tmpNavigator->LocateGlobalPointAndSetup(vtx);
  if (!tmpVolume)
  {
    // Cannot find tmpVolume
    return vtxvol;
  }
```

However, the root cause is actually from an unchecked overlapping in detector construction.

The `WCBox` [placement](https://github.com/WCSim/WCSim/blob/a5c9d56908f39b0daedab41c674730d04ce11771/src/WCSimDetectorConstruction.cc#L408) is actually bugged.
An argument is missing from the original code
```
  G4VPhysicalVolume* physiWCBox = 
    new G4PVPlacement(transform,
		      logicWCBox,
		      "WCBox",
		      logicExpHall,
		      false,
		      checkOverlaps);
```
The correct one should be
```
  G4VPhysicalVolume* physiWCBox = 
    new G4PVPlacement(transform,
		      logicWCBox,
		      "WCBox",
		      logicExpHall,
		      false,
                      0,
		      checkOverlaps);
```
Now we see the overlap warning
```
Checking overlaps for volume WCBox ... 
-------- WWWW ------- G4Exception-START -------- WWWW -------
*** G4Exception : GeomVol1002
      issued by : G4PVPlacement::CheckOverlaps()
Overlap with mother volume !
          Overlap is detected for volume WCBox:0
          with its mother volume expHall
          at mother local point (3235.55,4936.43,1970.58), overlapping by at least: 21.6853 cm 
NOTE: Reached maximum fixed number -1- of overlaps reports for this volume !
*** This is just a warning message. ***
-------- WWWW -------- G4Exception-END --------- WWWW -------
```
The error should be solved by setting `expHallLength` probably
```
  G4double expHallLength = 3.*WCLength; //jl145 - extra space to simulate cosmic muons more easily

  G4cout << " expHallLength = " << expHallLength / m << G4endl;
  G4double expHallHalfLength = 0.5*expHallLength;

  if (expHallHalfLength<.5*WCLength+4.2*m) // modified length in WCSimDetectorConstruction::ConstructCylinder()
      expHallHalfLength=.5*WCLength+4.2*m;
```

If everybody is happy, I can make a PR to push the fix later.

**Comment by nickwp** [comment]
Thanks @kmtsui very much for this analysis and debugging. Sounds like a good solution, but is it possible to extract the size of `logicWCBox` instead of hard-coding `.5*WCLength+4.2*m` in case one of the other alternative methods to ConstructCyilinder is used (or if the ConstructCyilinder method chanages in future)?

For example, it looks like `ConstructRealisticPlacement` does things differently, returning a polyhedral volume that I think is length `config.MainWaterTankLength + 30*cm`. So the check of the size of `expHallHalfLength<.5*WCLength+4.2*m` may fail if `ConstructRealisticPlacement` is ever used to build a detector where `3.*WCLength < config.MainWaterTankLength + 30*cm`. I guess this specific example should never really be a problem unless the user has set `config.MainWaterTankLength` incorrectly.

So it would be nice if the check can work in the general case if that's possible, but maybe the fixed overlap check is enough?

**Comment by kmtsui** [comment]
@nickwp good point. I think using `G4VisExtent extent = logicWCBox->GetSolid()->GetExtent();` could give a more generalized way to extract the volume size

**Comment by nickwp** [comment]
fixed by #464
